### PR TITLE
[OTLPHTTP Exporter] OAuth2 Client Credentials Authorization support for Exporters using HTTP Clients

### PR DESCRIPTION
**Pull request by pavankrish123**
**Description:** 

1. Added support for exporters that use HTTP Clients to use OAuth2 Client credentials workflow (2-legged/machine-to-machine). 
2. When enabled as a part of HTTP Client Configuration, an HTTP Client with OAuth2 supported authorization transport is returned. 
3. The underneath [Go OAuth2](https://pkg.go.dev/golang.org/x/oauth2/clientcredentials) library does the token auto-refresh as necessary. 

Example 

```
exporters:
  otlphttp:
    endpoint: http://localhost:9000
    oauth2:
      client_id: someclientidentifier
      client_secret: someclientsecret
      token_url: https://autz.server/oauth2/default/v1/token
      scopes: ["some.resource.read"]
```

**Link to tracking Issue**:
#2785

**Testing:** : Added unit tests/functional tests. 

**Documentation:**  Added README for the config

### CODE REVIEW

**Comment by pavankrish123** [comment]
Note: Understand that we are chasing towards GA. But since OTLPHTTP exporter features is present in the [roadmap](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/roadmap.md) for the sake of completeness I am putting up this PR. If it is bit late please let me know happy to move the PR to draft and will resubmit it when it is a good time.

File: config/confighttp/confighttp_test.go
Comment (jpkrohling) [inline config/confighttp/confighttp_test.go:433]:
```suggestion
		Endpoint:   "https://example.com/v1/endpoint",
```

File: config/confighttp/confighttp_test.go
Comment (jpkrohling) [inline config/confighttp/confighttp_test.go:446]:
```suggestion
func TestOAuth2SettingsWithHeaders(t *testing.T) {
```

File: config/confighttp/confighttp_test.go
Comment (jpkrohling) [inline config/confighttp/confighttp_test.go:437]: This looks odd for two reasons: 

1. you'll end up with `https://some-server/v1/token/v1/token`, won't you?
2. both are fixed strings, you don't need Sprintf

File: config/configoauth2/configoauth2client.go
Comment (jpkrohling) [inline config/configoauth2/configoauth2client.go:22]: We use go-oidc in the `configauth` package. Do you know if we can use the same package in both places, or are they complimentary to each other?

File: config/configoauth2/configoauth2client_test.go
Comment (jpkrohling) [inline config/configoauth2/configoauth2client_test.go:36]:
```suggestion
				TokenURL:     "https://example.com/v1/token",
```

**Review by jpkrohling — COMMENTED** [review]
We have two places with related code:
- https://github.com/open-telemetry/opentelemetry-collector/tree/main/config/configauth
- https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/configgrpc/bearer_token.go

This second seems to be closely related to this but won't perform token refreshes. They should really be made consistent among each other, either by having this here only taking care of sending the token that was previously configured, or by making that one to refresh tokens (and possibly use the same configuration as this here). The use case that the gRPC one covers is for Kubernetes deployments, where a token file is available locally in a file.

File: config/confighttp/confighttp_test.go
Comment (pavankrish123) [reply config/confighttp/confighttp_test.go]: Fixed

File: config/configoauth2/configoauth2client.go
Comment (pavankrish123) [reply config/configoauth2/configoauth2client.go:22]: That is correct, these packages are complimentary, go-odic is built on x/oauth2 handling ODIC (which itself is a specification on top of OAuth2) specific tasks

File: config/configoauth2/configoauth2client_test.go
Comment (pavankrish123) [reply config/configoauth2/configoauth2client_test.go:36]: fixed.

File: config/confighttp/confighttp_test.go
Comment (pavankrish123) [reply config/confighttp/confighttp_test.go:437]: Aha! sorry embarrassing copy paste error fixed now.

**Comment by pavankrish123** [comment]
@jpkrohling agreed. Let me refresh this PR after making the modules consistent. I will move `bearer_token.go`  into `configclientauth` package which has both `RoundTripper for OAuth2 (HTTP  Clients)`, `OAuth2PerRPCCredentials (grpc)` and `tokenRPCCredentials(grpc)` existing code.

**Comment by pavankrish123** [comment]
Hi @jpkrohling couple of thoughts here.

> This second seems to be closely related to this but won't perform token refreshes. They should really be made consistent among each other, either by having this here only taking care of sending the token that was previously configured, or by making that one to refresh tokens (and possibly use the same configuration as this here).

So as I understand we are looking into two completely different ways for authorization.

1. "bearer token" :  As I understand, we are directly taking the token through the configuration file and plugging in to headers.  This might be a viable option for long lived tokens in some use cases where security is not a concern. Clients in this scenario are directly dealing with tokens.  To support this flow
    - for grpc clients `bearer_token.go` is handling it by creating PerRPCCredentials to plug in the authorization header into grpc calls.
    - for http clients: we can plug in the `authorization: Bearer <token>` via `headers` option in `HTTPClientSettings` in `confighttp.go` - so no additional support is needed. 

2. OAuth2 Client Credential Flow:  [in this flow](https://tools.ietf.org/html/rfc6749#section-4.4) there is no token directly configured or entered into configuration file. The client (collector) authenticates server using client credentials grant and the tokens are received and refreshed as needed. Clients only deal with  <client-id , client-secret, token-url >  as opposed to token directly in this flow.  Token is derived indirectly and always lives in memory and need not be entered into files. The use cases for this scenario are

1.    we do not want to enter the access token directly into configuration file for safety reasons 
2.    access tokens are usually short lived and constant manual recycling is not an option. 

This is the flow the PR is aiming to support and so

-  for gRPC clients, `OAuth2ClientCredentialsPerRPCCredentials` function in `configauthclient/configoauth2.go` that produces  PerRPCCredentials  from an OAuth2 token source which gets the access token and auto refreshes it as needed.
-  for http clients:    `OAuth2RoundTripper` function in configauthclient/configoauth2.go` that produces a transport that manages OAuth2 tokens (getting, auto refreshing)

IMHO there is a validity in both the flows and the support should kept as it is.  Kindly let me know if I have mistaken somewhere and assumed something wrong. 

Notes:

- I have moved bearer_token.go to `configauthclient` module from `configgrpc` to make it live with related files that deal with client side authorization. 
- I have not enabled "client credentials" flow in configgrpc. Plan to do it in another PR. I would want to limit the scope for http clients in this one. Kindly let me know if this is alright. Happy to open an issue. 

Thanks for the review

**Comment by jpkrohling** [comment]
Your assessment is correct, there's one aspect that is missing from the current bearer token auth though: I originally proposed reading the token from a file, like the ones Kubernetes injects into pods. When those tokens are changed, the collector would reload them. Reading values from files were deemed as a feature that the collector should support more generically, so, I removed this from the original PR.

In that case, the OAuth dance would be done by an external process, and the current token to be used would be made ready to the collector.

I'll take another look at your PR soon.

File: config/configclientauth/README.md
Comment (jpkrohling) [inline config/configclientauth/README.md:3]:
```suggestion
This module provides helpers for HTTP/gRPC exporters, allowing them to be configured to perform authentication of outgoing requests.
```

File: config/configclientauth/README.md
Comment (jpkrohling) [inline config/configclientauth/README.md:1]:
```suggestion
# Authentication Settings for Exporters
```

File: config/configclientauth/README.md
Comment (jpkrohling) [inline config/configclientauth/README.md:11]:
```suggestion
This should be used when the client (exporter) is acting on its own behalf.
```

File: config/configclientauth/README.md
Comment (jpkrohling) [inline config/configclientauth/README.md:14]:
```suggestion
When enabled as a part of an HTTP/gRPC exporter configuration, the corresponding Client with OAuth2 supported authorization transport is returned.
```

File: config/configclientauth/README.md
Comment (jpkrohling) [reply config/configclientauth/README.md:1]: Correct me if I'm wrong, but from the client's perspective, this is only about authentication, not authorization.

File: config/configclientauth/README.md
Comment (jpkrohling) [inline config/configclientauth/README.md:20]:
```suggestion
- `client_secret`: Registered client's secret.
```

Comment (jpkrohling) [reply config/configclientauth/README.md:20]: You might want to add some warnings like the ones from https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2341.

File: config/configclientauth/configoauth2.go
Comment (jpkrohling) [inline config/configclientauth/configoauth2.go:29]:
```suggestion
	errNoClientIDProvided     = errors.New("no ClientID provided in the OAuth2 exporter configuration")
```

The end-user is going to see this, and it should be very clear what this is about. When doing a manual test, make sure the exporter name is included as a field in the logs (should be automatically added already).

File: config/configclientauth/configoauth2_test.go
Comment (jpkrohling) [inline config/configclientauth/configoauth2_test.go:23]: Why do you need to rename this one?

File: config/confighttp/confighttp.go
Comment (jpkrohling) [inline config/confighttp/confighttp.go:53]:
```suggestion
	// OAuth2Settings to use when performing client authentication against a remote endpoint
```

**Review by jpkrohling — COMMENTED** [review]
Looks nice! As a follow-up PR, we might want to create a doc somewhere with the auth information for both receivers and exporters.

File: config/configclientauth/README.md
Comment (pavankrish123) [reply config/configclientauth/README.md:1]: @jpkrohling this is interesting - yes, it is indeed client authentication but this is done inorder to get a token used for authorization to access resources on the server - not sure how to frame this.

File: config/configclientauth/README.md
Comment (pavankrish123) [reply config/configclientauth/README.md:3]: Hmm, makes me think - The ultimate goal in both OAuth2 client credentials and "Bearer Token" way is to get a token for authorization to access resources on the server - not sure how to go about it.

File: config/configclientauth/configoauth2_test.go
Comment (pavankrish123) [reply config/configclientauth/configoauth2_test.go:23]: just for clarity between `oauth2.<>` (golang oauth library) and `oauth.<>` (grpc library)

**Comment by pavankrish123** [comment]
Thanks for all the suggestions @jpkrohling  - I will add the warnings suggestions and refresh the PR and will open issues to document the auth receivers and exporters

File: config/configclientauth/README.md
Comment (pavankrish123) [reply config/configclientauth/README.md:3]:
```
# AuthX Settings for Exporters

This module provides HTTP/GRPC client types to be configured to perform authorization/authentication for requests using the following 
```
would something like this work?

Comment (jpkrohling) [reply config/configclientauth/README.md]: I would still use "HTTP/gRPC exporters" instead of client types, to make it extra clear that this is affecting exporters, not extensions nor receivers. But despite the name of the HTTP header being "authorization", sending credentials is only the authentication part of the story: it's telling the server who the client is (username), and giving some proof of it (secret/password). Authorization (is the user authorizes to perform the requested operation?) is solely done on the server-side.

Comment (pavankrish123) [reply config/configclientauth/README.md]: @jpkrohling got it! I have modified the README per your original suggestion. Please let me know - Thanks for the review!

**Review by jpkrohling — APPROVED** [review]

**Review by gramidt — APPROVED** [review]

**Comment by jpkrohling** [comment]
@tigrannajaryan, @bogdandrutu, this is ready to be merged from my perspective.

**Comment by tigrannajaryan** [comment]
I will not have time to review this, I am working on GA related tasks. @bogdandrutu I am not sure if you are able to review.

If Bogdan is also not able to review we may need to hold this for a while.

@pavankrish123 to avoid being blocked please follow the contributing guides in the future (we ask to submit an issue and discuss with maintainers before submitting a PR).

>If you would like to work on something that is not listed as an issue (e.g. a new feature or enhancement) please first read our vision and roadmap to make sure your proposal aligns with the goals of the Collector, then create an issue and describe your proposal. It is best to do this in advance so that maintainers can decide if the proposal is a good fit for this repository. This will help avoid situations when you spend significant time on something that maintainers may decide this repo is not the right place for.

**Comment by pavankrish123** [comment]
@tigrannajaryan / @bogdandrutu   thanks and I completely understand.  Will make sure to open an issue/proposal moving forward. I have assumed since otlphttp exporter is part of the roadmap - auth would be a necessary thing. 

Happy to check back once things settle down! Please let me know if I need to do anything in the meantime 

Pavan

**Comment by tigrannajaryan** [comment]
I believe we can keep this PR open and look at it when we are done with the GA. Please convert it to draft PR to avoid being marked as stale.

**Comment by pavankrish123** [comment]
Will do. Thanks @tigrannajaryan

**Comment by pavankrish123** [comment]
cc: @alolita just fyi

**Comment by pavankrish123** [comment]
Note: Can be modified based on how #2229 is concluded. Probably have to resubmit a new PR. I will hold off on this till the resolutions comes through

**Comment by pavankrish123** [comment]
Please hold this PR til #2229 is resolved. Based on the resolution, the configuration might change.

**Comment by pavankrish123** [comment]
Hi @jpkrohling @tigrannajaryan - the PR is marked stale and is due for deletion in couple of days despite the fact that we put it in draft mode. Any suggestions? I am OK reissuing the PR at the proper time as well.

Thanks Pavan

**Comment by jpkrohling** [comment]
There's a bigger discussion going on as part of #2603

**Comment by pavankrish123** [comment]
@jpkrohling Think the PR is auto closed despite it is a draft. Shall I reissue a fresh PR after  #2603 - I do not know how to reopen this one :(

**Comment by jpkrohling** [comment]
@tigrannajaryan, @bogdandrutu please reopen this one, as this was closed despite being in draft mode.

**Comment by bogdandrutu** [comment]
It says that there is already an open PR from the same branch

**Comment by pavankrish123** [comment]
Hello @bogdandrutu @jpkrohling Is there any chance we could get this thing reviewed now that @jpkrohling's #2603 is almost done. This PR is sort of independent to the work going on there but still belongs to Auth bucket we are tracking - This PR  is to expose OAuth2 configuration for clients (oltp http exporter) authentication to OAuth2 servers. This is different from 2603 which handles server (receiver) side Auth.

I believe @gramidt  is also looking for this something like this PR but on gRPC side. If we can get this through in upcoming release, we can start messaging OT collector to our customers (internal and external) to start using our solution - Thank you!

**Comment by jpkrohling** [comment]
>This PR is to expose OAuth2 configuration for clients (oltp http exporter) authentication to OAuth2 servers

Should this follow the same pattern as the server-side mechanism, where the clients would determine what to do based on a config option and performed by an extension?

**Pull request closed without merging**
